### ISSUE TITLE
test/python/test_dagcircuit.py has no assert

### ISSUE THREAD

**Issue by 1ucian0** [Qiskit/qiskit-terra#305]
The test `vi test/python/test_dagcircuit.py` is too basic and contain no assert.

## Expected Behavior
The tests in the testsuite should test something.

## Current Behavior
Right now, the test only implicit test for this file is "there is no exception".

## Possible Solution
Add asserts for the expected structures.

**Comment by eddieschoute** [comment]
Some use cases for the DAGCircuit.

1. Investigating which operations are being performed in the circuit:
    ```
    node_data = map(lambda n: n[1], dag.multi_graph.nodes(data=True))
    op_nodes = list(filter(lambda n: n["type"] == "op", node_data))
    ```
1. 2-qubit operations and ignoring barriers
    ```
    op_nodes = list(filter(lambda n: n["type"] == "op", node_data))
    binop_nodes = filter(lambda n: len(n["qargs"]) == 2 and n["name"] != "barrier", op_nodes)
    ```
1. Inspecting qargs with `node["qargs"]` to find out which qubits are used in an operation.
1. Given a list of swaps (`Tuple[_V,_V]`), produce a DAGCircuit that implements the swaps. The fact that the input type `_V` does not directly map to a DAGCircuit qubit necessitates the `inputmap` dictionary, that maps `_V` to DAG qubits.
    ```python
    def circuit(swaps: List[List[Swap[_V]]], allow_swaps=False) \
            -> Tuple[DAGCircuit, Dict[_V, Tuple[str, int]]]:
        """Produce a circuit description of a list of swaps.
    
            With a given permutation and permuter you can compute the swaps using the permuter function
            then feed it into this circuit function to obtain a circuit description.
    
            :param swaps: A list of swaps to perform.
            :param allow_swaps: Will use "cx" in basis of circuit by default,
                but if swaps are allowed, will use those instead.
            :return: A tuple with the circuit and a mapping of node to qubit in the circuit.
        """
        # Construct a circuit with each unique node id becoming a quantum register of size 1.
        dag = DAGCircuit()
        dag.add_basis_element('cx', 2)
        if allow_swaps:
            dag.add_basis_element('swap', 2)
    
        nodes = {
            swap_node
            for swap_step in swaps
            for swap_nodes in swap_step
            for swap_node in swap_nodes
        }
    
        inputmap: Dict[_V, Tuple[str, int]] = {}
    
        for node in nodes:
            dag.add_qreg(str(node), 1)
            inputmap[node] = (str(node), 0)
    
        def swap(node0, node1):
            """A circuit implementing a SWAP in the basis."""
            if "swap" in dag.basis:
                return [("swap", [node0, node1])]
    
            return [
                ("cx", [node0, node1]),
                ("cx", [node1, node0]),
                ("cx", [node0, node1])
            ]
    
        # Apply swaps to the circuit.
        for swap_step in swaps:
            for swap0, swap1 in swap_step:
                for gate in swap((str(swap0), 0), (str(swap1), 0)):
                    dag.apply_operation_back(*gate)
    
        return dag, inputmap

    ````
    I would also like to note that the whole checking if "swap" is in the basis seems like it should not be necessary. Maybe by defining a gate that can decompose, if necessary, to c-nots. I have not figured out how to use the `DAGCircuit.add_gate_data` to do anything useful.
1. Subsequently to making a DAGCircuit which implements the swaps I append it to my existing DAG (`arch_circuit`):
    ```
    wire_map = {k: k for k in dag.output_map}
    arch_circuit.compose_back(dag, wire_map=wire_map)
    ```
   It would also be great if we could compose back without an explicit wire_map. It could then just assume that each wire is mapped to a similarly-named wire on the other dag. And each unused wire would then be an identity circuit.

   Note: By inspection I found out the above code is wrong but runs without a hiccup and adds a `q` quantum register of size 3 (as defined by some other code) to `arch_circuit`. The `arch_circuit` at first only contains registers `(i,0)` for i∈[3]. Thus it is not checking the wire_map properly.

**Comment by eddieschoute** [comment]
I would like to note that the documentation of `compose_back` is ambiguous in the explanation of `wire_map`. What type does the dict have? Does it map `input_circuit` qubits to `self` or vice-versa?

Corrected code from item 5 above:
```python
wire_map = {
    _circuit_name(node): qubit
    for node, qubit in mapping_circuit.inputmap.items()
    }
arch_circuit.compose_back(mapping_circuit.circuit, wire_map=wire_map)
```
where the mapping circuit and inputmap are produced by the circuit function at item 4 above. And `_circuit_name` is a function I need to convert my node ids to safe quantum register names:
```python
def _circuit_name(graph_node: int) -> Tuple[str, int]:
    """Map a node id to a the circuit's qubit id."""
    return str(graph_node), 0
```

**Comment by ajavadia** [comment]
This is fixed in #1043.
